### PR TITLE
Cosmos spark3.3 support

### PR DESCRIPTION
**Pull request by TheovanKraay**
# Description

Updates required to support Spark 3.3 for Cosmos DB Spark Connector

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: .vscode/cspell.json
Comment (kushagraThapar) [inline .vscode/cspell.json:277]: I am not sure why we would need these 3 values here. All of them seem like command line tools for building the package. 

May be we do, I am not that familiar with vscode, however, if these were working without these 3 commands, then I would expect it to continue working without the need to add these :)

File: eng/pipelines/templates/stages/cosmos-emulator-matrix.json
Comment (kushagraThapar) [inline eng/pipelines/templates/stages/cosmos-emulator-matrix.json:58]: FYI, I thought for spark 3.*, we will need Java > 11. 
@xinlian12 worth checking on this before we merge this PR.

File: eng/versioning/version_client.txt
Comment (kushagraThapar) [inline eng/versioning/version_client.txt:96]: Should we start at 4.14.1 release? Also, there is no release for this package as 4.14.1 so I guess things might fail once we release this for the first time. 
cc @alzimmermsft  how do you propose we add this new module with new release versions?

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md:31]: is the min java version 8 or 11 for spark 3.3?

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosCatalog.scala
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosCatalog.scala:14]: just for my curiosity, why do we need to add CosmosCatalog to both spark 3-1 and 3-2?

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosCatalog.scala:14]: Nevermind, just saw the movement of CosmosCatalog to all three modules now, thanks!

**Review by kushagraThapar — COMMENTED** [review]
LGTM, great work @TheovanKraay  and @xinlian12  for the help. 

@alzimmermsft  - since you worked on cosmos spark modules before, can you please verify the build / pom.xml changes that we have done here. 
Also, we found out that all tests were running multiple times against all spark versions. Tried fixing it by moving test profiles to specific pom.xmls for spark 3-1, 3-2 and 3-3. Please review them and share your thoughts!

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md
Comment (alzimmermsft) [inline sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md:100]: This looks to be using the wrong version for supporting Databricks Runtime 8 as the table above says only `com.azure.cosmos.spark:azure-cosmos-spark_3-1_2-12:4.14.1` supports Runtime 8

**Review by alzimmermsft — APPROVED** [review]

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md:100]: Yeah, agreed, I think we need to fix this.

File: eng/versioning/version_client.txt
Comment (kushagraThapar) [reply eng/versioning/version_client.txt:96]: Good to know, I think we would better be consistent rather than starting a new version completely, I will update it.

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md]: Fixed.

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md
Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md]: @TheovanKraay  can you please confirm this?

File: eng/versioning/version_client.txt
Comment (kushagraThapar) [reply eng/versioning/version_client.txt]: fixed.

File: eng/pipelines/templates/stages/cosmos-emulator-matrix.json
Comment (kushagraThapar) [reply eng/pipelines/templates/stages/cosmos-emulator-matrix.json:58]: fixed.

File: .vscode/cspell.json
Comment (kushagraThapar) [reply .vscode/cspell.json:277]: @TheovanKraay  I think we should revert this change, but not sure.

Comment (TheovanKraay) [reply .vscode/cspell.json:277]: One of the checks fails without it... it is using this file to check for spelling errors in markdown files I believe.

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md]: I cannot find clear information. However, this doc https://spark.apache.org/docs/latest/ says:

_Spark runs on Java 8/11/17, Scala 2.12/2.13, Python 3.7+ and R 3.5+. **Java 8 prior to version 8u201 support is deprecated as of Spark 3.2.0**. When using the Scala API, it is necessary for applications to use the same version of Scala that Spark was compiled for. For example, when using Scala 2.13, use Spark compiled for 2.13, and compile code/applications for Scala 2.13 as well._

So, I think we should baseline our support at 11?

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md:13]: it exists for 3.1 and 3.2. Instead of the note, should we just start from fresh?

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/CONTRIBUTING.md
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3-3_2-12/CONTRIBUTING.md:13]: nit: should change to 3-3

File: sdk/cosmos/spark.yml
Comment (xinlian12) [inline sdk/cosmos/spark.yml:25]: working item to add databricks integration tests for 3.3: https://github.com/Azure/azure-sdk-for-java/issues/31834

**Review by xinlian12 — APPROVED** [review]
Look great, thanks :)

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md:14]: Do you mean prior to 4.15.0 ?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md:14]: Change log will be maintained for each Spark release version - so if we fix something it will show up in changelog for Spark 3.1, 3.2 and 3.3 - if 3.3 is introduced in 4.15 it is ok to refer to external files for < 4.15 - but starting 4.15 all changes need to be triplicated

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md:4]: Should be in section Feature Added - with link to the PR(s)

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md:14]: Makes sense, will change to refer to 4.15.0

File: eng/pipelines/templates/stages/cosmos-emulator-matrix.json
Comment (TheovanKraay) [reply eng/pipelines/templates/stages/cosmos-emulator-matrix.json:61]: makes sense, can we do that in subsequent PR when adding the databricks tests?

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosCatalog.scala
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosCatalog.scala:46]: This is functionally incorrect. You ignore the cascade flag. Our implementation always assumed cascading delete (so automatically deleted Coamos Database - which service-side will result in deleting containers). In Saprk 3.3 this should change - only if cascade==true we should call the base method. Otherwise throw an exception (not sure which one, needs some digging in Spark 3.3 spec)

This is blocking merge

File: sdk/cosmos/azure-cosmos-spark_3-2_2-12/README.md
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-2_2-12/README.md:53]: Are we planning to publish this beta to Maven? If not this should only be added when we actually publish a Spark 3.3 supporting version to Maven.

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md:18]: Should we link instead to Spark 3.2 links? They will be mostly compatible.

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md:95]: Should be version we actually publish to Maven

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosCatalog.scala
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosCatalog.scala:46]: Got it - so we should be if !cascade -> then check whether there is any containers under the database, if there is any, then throw some notEmpty exception, else just go ahead delete the database right?

File: sdk/cosmos/azure-cosmos-spark_3-2_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-2_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala:16]: NIT drop

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala:16]: NIT drop

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala:31]: NIT drop

File: sdk/cosmos/azure-cosmos-spark_3-2_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-2_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala:16]: NIT drop

File: sdk/cosmos/azure-cosmos-spark_3-2_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-2_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala:29]: NIT drop

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosCatalog.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosCatalog.scala:54]: LGTM

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala:18]: NIT drop

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala:33]: NIT drop

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala]: updated

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala]: updated

File: sdk/cosmos/azure-cosmos-spark_3-2_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3-2_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala]: updated

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala]: updated

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala]: updated

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md:18]: These seem to link to the base project. Not sure what your intention was here @FabianMeiswinkel?

File: sdk/cosmos/azure-cosmos-spark_3-2_2-12/README.md
Comment (TheovanKraay) [reply sdk/cosmos/azure-cosmos-spark_3-2_2-12/README.md:53]: @kushagraThapar your thoughts on this? I assume we will **not** publish to maven, so we should remove entry here until we do?

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-2_2-12/README.md:53]: I think we closed on publishing a beta to Maven - so your versions make sense.

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md:18]: Never mind - I meant the empty TODO with add more sections comments - but just saw they exist in Spark 3.2 and 3.1 README as well :-). So, ignore this comment.

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/README.md]: Yes - we will publish a beta to Maven

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md]: Fixed

File: sdk/cosmos/azure-cosmos-spark_3-2_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-2_2-12/src/test/scala/com.azure.cosmos.spark/CosmosCatalogITest.scala]: Fixed

File: sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3-3_2-12/CHANGELOG.md]: Fixed

**Pull request merged**
